### PR TITLE
DB-9665 Clear ZNode for OlapServer app watcher

### PR DESCRIPTION
**Pull request by dgomezferro**
Restart OlapServer if software version changed

### CODE REVIEW

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 1]
```diff
@@ -244,6 +244,24 @@ public void run() {
 
     }
 
+    private byte[] getKeepAlivePayload() {
+        DatabaseVersion version;
+        try {
+            version = JMXUtils.getLocalMBeanProxy(JMXUtils.SPLICEMACHINE_VERSION, DatabaseVersion.class);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
```
Comment (hatyo) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:253]: Is it needed to wrap the exception in a `RuntimeException`, can we just remove the catch?

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 2]
```diff
@@ -244,6 +244,24 @@ public void run() {
 
     }
 
+    private byte[] getKeepAlivePayload() {
+        DatabaseVersion version;
```
Comment (OlegMazurov) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:248]: Does _version_ need to be re-read on every call? Can it be an instance, or even static, variable similar to _OlapServerMaster_?

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 3]
```diff
@@ -166,10 +163,13 @@ public void run() {
...
+                            String keepAlivePath = root + HBaseConfiguration.OLAP_SERVER_PATH + HBaseConfiguration.OLAP_SERVER_KEEP_ALIVE_PATH + "/" + appName;
+
+                            byte[] payload = getKeepAlivePayload();
                             try {
-                                zk.create(path, payload, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.EPHEMERAL);
+                                // clean up node if it exists
+                                ZkUtils.safeDelete(keepAlivePath, -1);
+                                zk.create(keepAlivePath, payload, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.EPHEMERAL);
                             } catch (KeeperException ke) {
                                 if (ke.code().equals(NODEEXISTS)) {
```
Comment (OlegMazurov) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:174]: Is this code possible after the change? Should we still ignore it?

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java (hunk 1 above)
Comment (dgomezferro) [reply hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java]: I removed that as part of the refactoring to address other comments

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java (hunk 2 above)
Comment (dgomezferro) [reply hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java]: I initialize a keepAlive message prototype once in the beginning with the version information and then use that to generate the payload with the time, so we only access the version once per OlapServer lifetime.

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java (hunk 3 above)
Comment (dgomezferro) [reply hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java]: I removed it, good point

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**